### PR TITLE
WIP allow airtime transfer through transferTo

### PR DESCRIPTION
**Pull request by norkans7**

### CODE REVIEW

**Comment by norkans7** [comment]
Please review

File: temba/events/models.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+                                help_text="The channel that this event is relating to")
+    recipient = models.CharField(max_length=64)
+    amount = models.FloatField()
+    denomination = models.CharField(max_length=32, null=True, blank=True)
+    transaction_id = models.CharField(max_length=256, null=True, blank=True)
+    reference_operator = models.CharField(max_length=64, null=True, blank=True)
+    airtime_amount = models.CharField(max_length=32, null=True, blank=True)
+    last_message = models.CharField(max_length=256, null=True, blank=True)
+    data = models.TextField(null=True, blank=True)
+    event_log = models.TextField(null=True, blank=True)
```
Comment (nicpottier) [inline temba/events/models.py]: I would reduce the number of fields here to be only the ones we populate. IE, org, status, channel, recipient, amount, and denomination. I would then add a field for the raw request body (as text) and response (as text). That'll give us a lot more flexibility in logging if we ever plug in another provider, plus we really don't need to query on any of these other fields I don't think.

File: temba/events/migrations/0001_initial.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
```
Comment (nicpottier) [inline temba/events/migrations/0001_initial.py]: Why `events` as a package as opposed to `airtime`? That seems more appropriate.

File: temba/events/models.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+    @classmethod
+    def post_transferto_api_response(cls, login, token, **kwargs):
+        if not settings.SEND_WEBHOOKS:
```
Comment (nicpottier) [inline temba/events/models.py]: Probably makes sense to add a new variable for this. `SEND_AIRTIME`

File: temba/events/models.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+        data = kwargs
+        data.update(dict(login=login, key=key, md5=md5))
+
+        response = requests.post(cls.TRANSFERTO_AIRTIME_API_URL, data)
+
+        return response
+
+    @classmethod
+    def translate_transferto_response_content_as_json(cls, content):
```
Comment (nicpottier) [inline temba/events/models.py]: This looks like it is just URL encoded?

File: temba/events/models.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+        # get the results for this contact
+        results = flow.get_results(contact)
+        values = []
+
+        if results and results[0]:
+            values = results[0]['values']
+            for value in values:
+                value['time'] = datetime_to_str(value['time'])
+                value['value'] = unicode(value['value'])
```
Comment (nicpottier) [inline temba/events/models.py]: Seems like the value should be passed into `trigger_flow_event`.. airtime lib shouldn't have to know a lot about runs and flows, rather the caller should take care of that. (the action)

Comment (nicpottier) [inline temba/events/models.py]: So I would expect the signature here to be something like:
  `trigger_airtime_transfer(org, step, urn, denomination)`

File: temba/events/models.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        airtime_event = AirtimeEvent.objects.create(org=org, channel=channel, recipient=contact_urn.path,
+                                                    amount=0, data=json.dumps(data), event_log='',
+                                                    created_by=api_user,
+                                                    modified_by=api_user)
+
+        message = "None"
+        try:
+            action = 'msisdn_info'
+            request_kwargs = dict(action=action, destination_msisdn=airtime_event.recipient)
+            status_code, content_json, content = airtime_event.get_transferto_response_json(**request_kwargs)
```
Comment (nicpottier) [inline temba/events/models.py]: Naming here is a bit goofy.. you aren't converting to json, you are converting to a dict. You can probably simplify this to something like `parse_transferto_response`

File: temba/events/models.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            amount_config = ruleset.config_json()
+            country_config = amount_config.get(country_code, dict())
+            amount = country_config.get('amount', 0)
+
+            data['transferto_dumps'][action] = dict(status_code=status_code, data=content_json)
+            airtime_event.amount = amount
+            airtime_event.data = json.dumps(data)
+            airtime_event.event_log += content + AirtimeEvent.LOG_DIVIDER
+
+            product_list = content_json.get('product_list', [])
```
Comment (nicpottier) [inline temba/events/models.py]: Seems like doing two calls for every single transfer might be a bit wasteful. Wonder if it makes sense to cache the possible denominations in Redis, maybe with a one hour timeout. Could add refetching this if we get an error from TransferTo about a missing denomination when trying to initiate a transfer.

File: temba/events/models.py (hunk 5 above)
Comment (nicpottier) [inline temba/events/models.py]: This is a pretty massive method with lots going on.. would be good to break it up, which will make testing it simpler and a lot easier to read. Add some comments while you are at it.

File: temba/flows/views.py [hunk 8]
```diff
@@ -1305,10 +1305,13 @@ def get(self, request, *args, **kwargs):
 
             # all the translation languages for our org
             languages = [lang.as_json() for lang in flow.org.languages.all().order_by('orgs')]
-
+            org_channel_countries = flow.org.get_org_channel_countries()
```
Comment (nicpottier) [inline temba/flows/views.py]: For consistency maybe just `channel_countries`

File: temba/orgs/models.py [hunk 9]
```diff
@@ -512,6 +515,20 @@ def get_webhook_headers(self):
         """
         return json.loads(self.webhook).get('headers', dict()) if self.webhook else dict()
 
+    def get_org_channel_countries(self):
+        org_channel_countries = []
+        channels_countries = self.channels.filter(is_active=True).exclude(country=None).values('country').distinct()
```
Comment (nicpottier) [inline temba/orgs/models.py]: might be able to simplify this slightly by using `values_list('country', flat=True)`

File: temba/orgs/models.py [hunk 10]
```diff
@@ -551,6 +568,39 @@ def trigger_send(self, msgs=None):
...
+    def connect_transferto(self, account_login, airtime_api_token):
+        transferto_config = {TRANSFERTO_ACCOUNT_LOGIN: account_login.strip(),
+                             TRANSFERTO_AIRTIME_API_TOKEN: airtime_api_token.strip()}
+
+        config = self.config_json()
+        config.update(transferto_config)
+        self.config = json.dumps(config)
+
+        # clear all our channel configurations
+        self.save(update_fields=['config'])
```
Comment (nicpottier) [inline temba/orgs/models.py]: Do we need this?

File: temba/orgs/models.py [hunk 11]
```diff
@@ -551,6 +568,39 @@ def trigger_send(self, msgs=None):
...
+
+    def is_connected_to_transferto(self):
```
Comment (nicpottier) [inline temba/orgs/models.py]: passing 'user' to these and updating `modified_on` and `modified_by` would be best practice.

File: temba/orgs/views.py [hunk 12]
```diff
@@ -1648,6 +1651,76 @@ def derive_formax_sections(self, formax, context):
...
+                super(OrgCRUDL.TransfertoAccount.TransfertoAccountForm, self).clean()
+                if self.cleaned_data.get('disconnect', 'false') == 'false':
+                    account_login = self.cleaned_data.get('account_login', None)
+                    airtime_api_token = self.cleaned_data.get('airtime_api_token', None)
+
+                    try:
+                        from temba.events.models import AirtimeEvent
+                        response = AirtimeEvent.post_transferto_api_response(account_login, airtime_api_token, action='ping')
+                    except:
+                        raise ValidationError(_("Your TransferTo API key and secret seem invalid. "
```
Comment (nicpottier) [inline temba/orgs/views.py]: Do we have more details from TransferTo we can include in this validation error?

File: temba/utils/__init__.py [hunk 13]
```diff
@@ -42,6 +43,13 @@
 }
 
 
+MISSING_COUNTRY_CODE_NAME_MAP = {
+    'Democratic Republic of the Congo': 'CD',
+    'Ivory Coast': 'CI',
+    'United States': 'US',
+}
```
Comment (nicpottier) [inline temba/utils/__init__.py]: Where did these come from?

File: templates/partials/split_editor.haml [hunk 14]
```diff
@@ -56,6 +56,19 @@
       #split-on
         -trans "If the keypad entry before the # symbol.."
 
+    .airtime{ng-if:'formData.rulesetConfig.type == "airtime"'}><
```
Comment (nicpottier) [inline templates/partials/split_editor.haml]: Is this conditional based on whether they have a TransferTo account connected?

**Comment by nicpottier** [comment]
Overall looks good, nice work! @ericnewcomer and @rowanseymour please take a look too, this is a big change.

File: temba/utils/__init__.py (hunk 13 above)
Comment (norkans7) [inline temba/utils/__init__.py]: TransferTo uses those as name and the libs we use have different names like "Cote d'Ivoire" I added these to map them correctly

File: templates/partials/split_editor.haml (hunk 14 above)
Comment (norkans7) [inline templates/partials/split_editor.haml]: What about they had the service connected and disconnect it? do we remove this again? Now it is always appearing

File: temba/events/migrations/0001_initial.py (hunk 2 above)
Comment (norkans7) [inline temba/events/migrations/0001_initial.py]: I'll change to airtime

File: temba/events/models.py (hunk 3 above)
Comment (norkans7) [inline temba/events/models.py]: yep

File: temba/events/models.py (hunk 4 above)
Comment (norkans7) [inline temba/events/models.py]: this is the parse_response

File: temba/events/models.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+            error_code = int(content_json.get('error_code', None))
+            error_txt = content_json.get('error_txt', None)
+
+            if error_code != 0:
+                message = "Got non-zero error code (%d) from TransferTo with message (%s)" % (error_code, error_txt)
+                airtime_event.status = AirtimeEvent.FAILED
+                raise Exception(message)
+
+            country_name = content_json.get('country', '')
+            country_code = get_country_code_by_name(country_name)
```
Comment (ericnewcomer) [inline temba/events/models.py]: There's no way to coerce TransferTo to give us country codes? Translating from names seems really fragile.

File: temba/utils/__init__.py (hunk 13 above)
Comment (ericnewcomer) [inline temba/utils/__init__.py]: Again, seems really fragile. If its necessary, maybe call it TRANSFERTO_COUNTRY_NAMES. Do you have an exhaustive list where these vary from our country library?

File: templates/partials/split_editor.haml (hunk 14 above)
Comment (ericnewcomer) [inline templates/partials/split_editor.haml]: I think we shouldn't give them option to add new ones if there's no account, but if the flow has one in it's definition we show it (but it's a noop). Same thing for importing into a non-transferto enabled account.

File: temba/events/models.py (hunk 15 above)
Comment (norkans7) [inline temba/events/models.py]: It would be awesome if they could adjust to include the country code for sure

**Comment by norkans7** [comment]
Replaced by #742

**Pull request closed without merging**